liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/DevMojo.java

```diff
@@ -809,12 +809,55 @@ public void checkConfigFile(File configFile, File serverDir) {
                             featureElems[i + 1] = element(name("feature"), values[i]);
                         }
-                        runLibertyMojoInstallFeature(element(name("features"), featureElems), super.getContainerName());
+                        if (generateFeatures) {
+                            runLibertyMojoGenerateFeatures(element(name("features"), featureElems));
+                        } else {
+                            runLibertyMojoInstallFeature(element(name("features"), featureElems),
+                                    super.getContainerName());
+                        }
                         this.existingFeatures.addAll(features);
+                    } else if (generateFeatures) {
+                        // TODO properly handle this case:
+                        // generate features even if the feature list has not changed
+                        // the user may explicitly specify a feature in their server.xml that was
+                        // included in the generated features file (the duplicate should be
+                        // removed from the generated file)
+                        runLibertyMojoGenerateFeatures(null);
                     }
                 }
             } catch (MojoExecutionException e) {
-                log.error("Failed to install features from configuration file", e);
+                if (generateFeatures) {
+                    log.error("Failed to install features from configuration file: " + configFile.getAbsolutePath()
+                            + ". Reverting the recent configuration file changes is recommended.", e);
+                } else {
+                    log.error("Failed to install features from configuration file: " + configFile.getAbsolutePath(), e);
+                }
+            }
+        }
+
+        @Override
+        public void generateAndUpdateFeatures(File serverDir) {
+            try {
+                runLibertyMojoGenerateFeatures(null);
+            } catch (MojoExecutionException e) {
+                if (e.getMessage().contains("feature conflict")) {
+                    log.error(
+                            "Failed to generate server features due to a feature conflict. Features will need to be manually configured.",
+                            e);
+                } else {
+                    log.error("Failed to generate server features, features will need to be manually configured.", e);
+                }
+
+                return;
+            }
+            ServerFeature servUtil = getServerFeatureUtil();
+            Set<String> features = servUtil.getServerFeatures(serverDir, libertyDirPropertyFiles);
+            if (features != null) {
+                features.removeAll(existingFeatures);
+                if (!features.isEmpty()) {
+                    this.existingFeatures.addAll(features);
+                    log.debug("Updating feature list: " + this.existingFeatures);
+                }
```

Comment:
It may help to add some comments on what this logic is doing.  

I realize that some of this is reused from `checkConfigFile()` above, but is there any need to do `features.removeAll(existingFeatures);`?  Can it just do `this.existingFeatures.addAll(features);` since existingFeatures is a Set so it would not have duplicates?

Reply:
If the features have not changed, `features.removeAll(existingFeatures)` will result in an empty list and we will not enter the if statement logic.

I've updated this method to match the one above, where if new features have been added, we display the message "Configuration features have been added".